components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/ServerURL.java

```diff
@@ -0,0 +1,411 @@
+
+import org.apache.axis2.engine.AxisConfiguration;
+import org.apache.commons.lang.StringUtils;
+import org.wso2.carbon.base.ServerConfiguration;
+import org.wso2.carbon.context.PrivilegedCarbonContext;
+import org.wso2.carbon.identity.core.URLResolverException;
+import org.wso2.carbon.identity.core.internal.IdentityCoreServiceComponent;
+import org.wso2.carbon.identity.core.util.IdentityCoreConstants;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.utils.CarbonUtils;
+import org.wso2.carbon.utils.NetworkUtils;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+
+import java.net.SocketException;
+import java.util.ArrayList;
+
+import static org.wso2.carbon.identity.core.util.IdentityTenantUtil.isTenantQualifiedUrlsEnabled;
+
+/**
+ * Server URL representation.
+ */
+public class ServerURL {
+
+    private String protocol;
+    private String hostName;
+    private Integer port;
+    private String proxyContext;
+    private String webContext;
+    private Boolean addProxyContext;
+    private Boolean addWebContext;
+    private String urlContext;
+    private Boolean appendTenantAsQueryParam;
+    private Boolean appendTenantAsPathParam;
+    private ArrayList<String> parameters;
+    private ArrayList<String> fragments;
+
+    private ServerURL(ServerURLBuilder serverURLBuilder) {
+
+        this.protocol = serverURLBuilder.protocol;
+        this.hostName = serverURLBuilder.hostName;
+        this.port = serverURLBuilder.port;
+        this.proxyContext = serverURLBuilder.proxyContext;
+        this.webContext = serverURLBuilder.webContext;
+        this.addProxyContext = serverURLBuilder.addProxyContext;
+        this.addWebContext = serverURLBuilder.addWebContext;
+        this.urlContext = serverURLBuilder.urlContext;
+        this.appendTenantAsQueryParam = serverURLBuilder.appendTenantAsQueryParam;
+        this.appendTenantAsPathParam = serverURLBuilder.appendTenantAsPathParam;
+        this.parameters = serverURLBuilder.parameters;
+        this.fragments = serverURLBuilder.fragments;
+    }
+
+    public String getProtocol() {
+
+        return protocol;
+    }
+
+    public String getHostName() {
+
+        return hostName;
+    }
+
+    public Integer getPort() {
+
+        return port;
+    }
+
+    public String getProxyContext() {
+
+        return proxyContext;
+    }
+
+    public String getWebContext() {
+
+        return webContext;
+    }
+
+    public String getUrlContext() {
+
+        return urlContext;
+    }
+
+    public Boolean getAddProxyContext() {
+
+        return addProxyContext;
+    }
+
+    public Boolean getAddWebContext() {
+
+        return addWebContext;
+    }
+
+    public Boolean getAppendTenantAsQueryParam() {
+
+        return appendTenantAsQueryParam;
+    }
+
+    public Boolean getAppendTenantAsPathParam() {
+
+        return appendTenantAsPathParam;
+    }
+
+    public ArrayList<String> getParameters() {
+
+        return parameters;
+    }
```

Comment:
Don't return the list which is mutable. Rather return as properties, immutable list, or better to provide the capability to get a specific parameter given a key